hash-tables/problem-set-hash-tables.md

```diff
@@ -0,0 +1,287 @@
+
+    def test_is_permutation_unmatching_lengths(self):
+        self.assertTrue(not is_permutation("pizza", "piza"))
+
+    def test_is_permutation_same_word(self):
+        self.assertTrue(is_permutation("pizza", "pizza"))
+
+    def test_is_permutation_empty_strings(self):
+        self.assertTrue(is_permutation("", ""))
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+def populate_frequency_map(text):
+    frequency_hash = {}
+    for char in text:
+        if frequency_hash.get(char):
+            frequency_hash[char] += 1
+        else:
+            frequency_hash[char] = 1
+    return frequency_hash
+
+
+def is_permutation(red, blue):
+    red_frequency_hash = populate_frequency_map(red)
+    blue_frequency_hash = populate_frequency_map(blue)
+
+    return red_frequency_hash == blue_frequency_hash
+```
+
+### !end-explanation
+### !end-challenge
+<!-- prettier-ignore-end -->
+
+<!-- Question 3 -->
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: 9LkdRV
+* title: Hash Tables
+* points: 3
+### !question
+
+Write a function `is_palindrome_permutation` that takes a string as an argument and returns `True` if the letters could be re-arranged into a palindrome.
+
+Here are the tests:
+
+```python
+def test_is_palindrome_permutation_simple():
+    assert is_palindrome_permutation("lool")
+
+
+def test_is_palindrome_permutation_palindrome():
+    assert is_palindrome_permutation("racecar")
+
+
+def test_is_palindrome_permutation_rearranged_palindrome():
+    assert is_palindrome_permutation("carrace")
+
+
+def test_is_palindrome_permutation_incomplete():
+    assert not is_palindrome_permutation("raceca")
+
+
+def test_is_palindrome_permutation_unmatching_letters():
+    assert not is_palindrome_permutation("hello")
+
+
+def test_is_palindrome_permutation_empty_string():
+    assert is_palindrome_permutation("")
+```
+
+### !end-question
+### !tests
+```python
+import unittest
+from main import *
+class TestChallenge(unittest.TestCase):
+    # Note: Written with these asserts to mimic the pytest assertions we give students
+
+    def test_is_palindrome_permutation_simple(self):
+        self.assertTrue(is_palindrome_permutation("lool"))
+
+    def test_is_palindrome_permutation_palindrome(self):
+        self.assertTrue(is_palindrome_permutation("racecar"))
+
+    def test_is_palindrome_permutation_rearranged_palindrome(self):
+        self.assertTrue(is_palindrome_permutation("carrace"))
+
+    def test_is_palindrome_permutation_incomplete(self):
+        self.assertTrue(not is_palindrome_permutation("raceca"))
+
+    def test_is_palindrome_permutation_unmatching_letters(self):
+        self.assertTrue(not is_palindrome_permutation("hello"))
+
+    def test_is_palindrome_permutation_empty_string(self):
+        self.assertTrue(is_palindrome_permutation(""))
+```
+### !end-tests
+### !explanation
```

Comment:
consider adding a hint as before to give the student some idea of _how_ to apply dictionaries to this problem.
```suggestion
### !hint
Many problems that use dictionaries to solve problems related to strings involve counting the available letters. What must be true about the counts of the letters in a string in order for it to be a palindrome? From the previous problem, can we say that the same must be true about a _permutation_ of a palindrome?
### !end-hint
### !explanation
```